src/overseer/store/jdbc.clj

```diff
@@ -0,0 +1,238 @@
+  {:mysql
+   {:table-spec "ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=DYNAMIC"}})
+
+(defn install' [adapter db-spec]
+  (j/db-do-commands
+    db-spec
+    [(j/create-table-ddl
+       :overseer_jobs
+       [[:id "VARCHAR(64) PRIMARY KEY"]
+        [:type "VARCHAR(255)"]
+        [:args "VARCHAR(2048)"]
+        [:status "VARCHAR(255)"]
+        [:failure "VARCHAR(2048)"]
+        [:heartbeat :int]
+        [:lock_version "INT NOT NULL DEFAULT '0'"]]
+       (get default-table-opts adapter {}))
+     (j/create-table-ddl
+       :overseer_dependencies
+       [[:job_id "VARCHAR(64)"]
+        [:dep_id "VARCHAR(64)"]]
+       (get default-table-opts adapter {}))
+     "CREATE INDEX index_overseer_jobs_on_status ON overseer_jobs (status)"
+     "CREATE INDEX index_overseer_dependencies_on_job_id ON overseer_dependencies (job_id)"
+     "CREATE INDEX index_overseer_dependencies_on_dep_id ON overseer_dependencies (dep_id)"])
+  :ok)
+
+(defn- jobs-with-unfinished-dep
+  "Return the seq subset of `job-ids` that have an unfinished direct dependency"
+  [db-spec job-ids]
+  (if (seq job-ids)
+    (->> (j/query
+           db-spec
+           (flatten [(str "SELECT job_id FROM overseer_dependencies "
+                          "INNER JOIN overseer_jobs ON overseer_jobs.id = overseer_dependencies.dep_id "
+                          (format "WHERE overseer_dependencies.job_id IN (%s) " (in-placeholders job-ids))
+                          "AND overseer_jobs.status != 'finished'")
+                     job-ids]))
+         (map :job_id))
+    []))
+
+(defmacro with-ignore-stale-objects [& body]
+  `(try
+    (do ~@body)
+    (catch clojure.lang.ExceptionInfo ex#
+      (when-not (= :stale-object (:type (ex-data ex#)))
+        (throw ex#)))))
+
+(defrecord JdbcStore [adapter db-spec]
+  core/Store
+  (install [this]
+    (install' adapter db-spec))
+
+  (transact-graph [this graph]
+    (assert (core/valid-graph? graph))
+    (let [job-rows (map job->jdbc-map (loom.graph/nodes graph))
+          dep-rows (job-dep-jdbc-maps graph)]
+      (j/with-db-transaction [t-con db-spec]
+        (j/insert-multi! t-con :overseer_jobs job-rows)
+        (j/insert-multi! t-con :overseer_dependencies dep-rows))))
+
+  (job-info [this job-id]
+    (query-job db-spec job-id))
+
+  (reserve-job [this job-id]
+    (let [where-map {:status "unstarted"}
+          set-map {:status "started"
+                   :heartbeat (core/heartbeat)}]
+      (with-ignore-stale-objects
+        (update-job db-spec job-id where-map set-map)
+        (query-job db-spec job-id))))
+
+  (finish-job [this job-id]
+    (let [where-map {:status "started"}
+          set-map {:status "finished"}]
+      (update-job db-spec job-id where-map set-map)))
+
+  (fail-job [this job-id failure]
+    (let [where-map {:status "started"}
+          set-map {:status "failed"
+                   :failure (pr-str failure)}]
+      (update-job db-spec job-id where-map set-map)))
```

Comment:
Yeah as we talked about it's really interesting to basically just watch this all model a state machine.  It makes me wonder whether the store should be composed of two protocols: one for listing stuff / transacting graphs, and a separate one that is just a very explicit finite state machine modeling a job.  You can see how finite state machine libraries often look:

https://github.com/cdorrat/reduce-fsm
https://github.com/ztellman/automat

It makes me think you could do the protocol like

```clojure
(def my-job (store/job store job-id))
;;=> #<JobAtom>

@my-job
;; => {:status :unstarted}

(core/start! my-job)
```

In other words there's a protocol defined on an atom-like value representing a job handle with atomic transitions/operations on it, as opposed to the protocol for the state transitions being defined on the store.